columnformers/models/layers.py

```diff
@@ -280,3 +283,240 @@ def init_weights(module: nn.Module):
+        bias: bool = True,
+        blocksize: int = 32,
+        in_shape: Literal["nlc", "nchw", "nd"] = "nchw",
+    ):
+        super().__init__()
+        assert isinstance(kernel_size, int), "only square kernels supported"
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.height = height
+        self.kernel_size = kernel_size
+        self.depthwise = depthwise
+        self.blocksize = blocksize
+        self.in_shape = in_shape
+        self.channels_last = not depthwise
+
+        connectivity = _sparse_local_connectivity(
+            in_channels,
+            out_channels,
+            kernel_size,
+            height,
+            depthwise=depthwise,
+            channels_last=self.channels_last,
+        )
+        self.bsl = BlockSparseLinear(
+            connectivity=connectivity, bias=bias, blocksize=blocksize
+        )
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        needs_reshape = self.in_shape != "nd"
+
+        if needs_reshape:
+            in_pattern = "n (h w) c" if self.in_shape == "nlc" else "n c h w"
+            out_pattern = "n (h w c)" if self.channels_last else "n (c h w)"
+            input = rearrange(
+                input, f"{in_pattern} -> {out_pattern}", h=self.height, w=self.height
+            )
+
+        output = self.bsl(input)
+
+        if needs_reshape:
+            output = rearrange(
+                output,
+                f"{out_pattern} -> {in_pattern}",
+                c=self.out_channels,
+                h=self.height,
+                w=self.height,
+            )
+        return output
+
+
+def _sparse_local_connectivity(
+    in_channels: int,
+    out_channels: int,
+    kernel_size: int,
+    height: int,
+    depthwise: bool = False,
+    channels_last: bool = False,
+    dtype: _dtype = None,
+    device: _device = None,
+) -> torch.Tensor:
+    """
+    Construct sparse local connectivity matrix, shape
+    (out_channels * height * height, in_channels * height * height). The returned
+    connectivity will have sparse COO layout.
+
+    The connectivity pattern is equivalent to
+    `nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding="same")`
+
+    If channels_last is True, the shape of connectivity in einops notation is
+    "(h w cout) (h w cin)". Otherwise, it is "(cout h w) (cin h w)". The latter should
+    be more efficient when depthwise is True (connectivity will be more block sparse).
+    """
+    assert kernel_size % 2 == 1, "kernel_size must be odd"
+    assert (
+        not depthwise or out_channels == in_channels
+    ), "in channels must match out channels for depthwise"
+    N = height * height
+
+    # ij indices of input grid
+    # (h^2, 2)
+    col_indices = torch.cartesian_prod(torch.arange(height), torch.arange(height))
+
+    # conv kernel index offsets. note that the kernel width is required to be odd.
+    # (k^2, 2)
+    kernel_half_width = (kernel_size - 1) // 2
```

Comment:
This is a bit restrictive, can we adapt this to also include even kernel size by doing something like this?

```
if kernel_size % 2 == 0:
    kernel_half_width = kernel_size // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width),
        torch.arange(-kernel_half_width, kernel_half_width),
    )
else:
    kernel_half_width = (kernel_size - 1) // 2
    kernel_indices = torch.cartesian_prod(
        torch.arange(-kernel_half_width, kernel_half_width + 1),
        torch.arange(-kernel_half_width, kernel_half_width + 1),
    )
```

Reply:
Yes I think you're right, something like this would probably be better. Although it feels like it should be possible to make the code shorter.

More generally, it would probably be best to have exactly the same interface and behavior as [Conv2d](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html). What I have now takes a few shortcuts.